experimental/sql/src/projection_pushdown.py

```diff
@@ -0,0 +1,86 @@
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+from dataclasses import dataclass
+from xdsl.ir import Operation, MLContext, Region, Block
+from typing import List, Type, Optional
+from xdsl.dialects.builtin import ArrayAttr, StringAttr, ModuleOp, IntegerAttr
+
+from xdsl.pattern_rewriter import (RewritePattern, GreedyRewritePatternApplier,
+                                   PatternRewriteWalker, PatternRewriter,
+                                   op_type_rewrite_pattern)
+
+import dialects.rel_alg as RelAlg
+
+
+@dataclass
+class ProjectionPushdownRewriter(RewritePattern):
+
+  def find_cols_in_expr(self, op: RelAlg.Expression) -> list[str]:
+    """
+    Returns the list of cols that are used in this operation or its children.
+    """
+    if isinstance(op, RelAlg.BinOp):
+      return self.find_cols_in_expr(op.lhs.op) + self.find_cols_in_expr(
+          op.rhs.op)
+    if isinstance(op, RelAlg.Column):
+      return [op.col_name.data]
+    if isinstance(op, RelAlg.Compare):
+      return self.find_cols_in_expr(op.left.op) + self.find_cols_in_expr(
+          op.right.op)
+    return []
+
+  @op_type_rewrite_pattern
+  def match_and_rewrite(self, op: RelAlg.Project, rewriter: PatternRewriter):
+    if not isinstance(op.input.op, RelAlg.Select):
+      return
+
+    select_cols = [
+        elem for sublist in (
+            self.find_cols_in_expr(o) for o in op.input.op.predicates.ops)
+        for elem in sublist
+    ]
+
+    project_cols = [
+        elem for sublist in (self.find_cols_in_expr(o)
+                             for o in op.projections.ops) for elem in sublist
+    ]
+
+    result_cols = [s.data for s in op.names.data]
+
+    # TODO: handle renames of columns
```

Comment:
In the spirit of my comment above, I think pushing a projection through a rename consists of (1) removing all renames where the new name does not occur in the projection, (2) apply the reverse rename to the projection (what happens if the rename is not injective, e.g., `a->b, a->c`?), and (3) swap the projection with the rename.